### PR TITLE
Optimized Python splitters

### PR DESCRIPTION
**Pull request by ChuyangKe**
### Description

Rewrote Python splitter functions to improve efficiency. 

Tested on synthetic datasets with 10 million records and 100,000 users. The runtime is reduced from 209.7s to 18.8s, indicating a 11x performance improvement.

### Checklist:

- [x] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.
- [x] This PR is being made to `staging branch` and not to `main branch`.

### CODE REVIEW

**Comment by miguelgfierro** [comment]
> DO NOT MERGE. More tests are needed.

A small notice. An alternative way to tell the team that you don't want us to merge is by creating a draft PR instead of a normal PR. Simon does it from time to time. There is no need to change this PR, this is just FYI in case you were not familiar :-)

File: recommenders/datasets/python_splitters.py [hunk 1]
```diff
@@ -87,36 +87,33 @@ def _do_stratification(
...
+        data = data.drop("random", axis=1)
 
-    return splits_list
+    splits = []
+    prev_threshold = None
+    for threshold in np.cumsum(ratio):
+        condition = data["rank"] <= round(threshold * data["count"])
+        if prev_threshold is not None:
+            condition &= data["rank"] > round(prev_threshold * data["count"])
+        splits.append(data[condition].drop("rank", axis=1).drop("count", axis=1))
```
Comment (miguelgfierro) [inline recommenders/datasets/python_splitters.py:113]:
```suggestion
        splits.append(data[condition].drop(["rank", "count"], axis=1))
```
Can you check if this change will get you a better time?

**Review by miguelgfierro — APPROVED** [review]
LGTM

Just added an idea to consider

File: recommenders/datasets/python_splitters.py (hunk 1 above)
Comment (ChuyangKe) [reply recommenders/datasets/python_splitters.py]: Yes, this saves an additional 10% of the time. Thanks!

**Pull request merged**